Add oc rsync command

csrwng opened the pull request:
Refactors existing rsh command to match upstream structure. Adds rsync command that will use 'rsync' if available and on Linux, otherwise will fall back to 'tar'.

In windows, it will use 'tar' and not try rsync.

csrwng commented:
@bparees @fabianofranz @smarterclayton

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,381 @@
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+	"gopkg.in/pipe.v2"
```
What is this used for?

csrwng replied:
To pipe tar output from local to remote exec

smarterclayton replied:
And os.Pipe can't do that for you?

csrwng replied:
The main reason was convenience. There was less setup required with this library. Not a big deal to go back to os.Pipe.

smarterclayton replied:
Would prefer not pulling in a pipe library if os.Pipe() works

On Thu, Sep 24, 2015 at 11:21 AM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/rsync.go
> <https://github.com/openshift/origin/pull/4787#discussion_r40331561>:
>
> > @@ -0,0 +1,381 @@
> > +package cmd
> > +
> > +import (
> > +	"bytes"
> > +	"fmt"
> > +	"io"
> > +	"os"
> > +	"os/exec"
> > +	"path/filepath"
> > +	"strings"
> > +
> > +	"github.com/golang/glog"
> > +	"github.com/spf13/cobra"
> > +	"gopkg.in/pipe.v2"
>
> The main reason was convenience. There was less setup required with this
> library. Not a big deal to go back to os.Pipe.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787/files#r40331561>.
>

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,381 @@
+
+This command will attempt to copy local files to a remote container. It will default to the
+first container if none is specified, and will attempt to use 'rsync' if available locally and in the
+container. If not present, it will attempt to use 'tar' to send files to the container.`
```
Why is tar a good fallback?

csrwng replied:
because if present, it can accept files from stdin in the container

smarterclayton replied:
Why am I likely to have tar if i don't have rsync?

csrwng replied:
After talking to @bparees at least locally you wouldn't need an external program. We can tar the directory for you in go and stream it into the pod. You'd only need the tar in the container.

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,381 @@
+	if len(o.ContainerName) > 0 {
+		rshCmd = append(rshCmd, "-c", o.ContainerName)
+	}
+	rshCmd = append(rshCmd, "sh", "-c", "'exec \"$@\"'")
```
Why isn't this using rsh?

csrwng replied:
It is using rsh:

```
rsync -av -e "oc rsh <podname> sh -c 'exec "\$@" ' " /tmp/source ignore:/tmp/dest
```

smarterclayton replied:
If you are using rsh, you don't need `sh -c exec ...` - that's what rsh *does*

smarterclayton replied:
And you shouldn't take pod name before the arg - you don't need it.  Rsh now works in rsync compatible form - `rsh <pod> <command>` so don't use `ignore`, use `podname`

csrwng replied:
I'm not using -i in the call to rsh, but are you talking about 'ignore:/' in the last parameter to rsync? If I don't have that there, rsync doesn't know that it's a remote destination.

smarterclayton replied:
The nested rsh command call you should be making is "oc rsh" - if you
generate the rsync command with podname:/foo instead of ignore:/foo, then
everything works correctly.  the exec @ thing is no longer necessary now
that rsh is fixed.

On Thu, Sep 24, 2015 at 11:18 AM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/rsync.go
> <https://github.com/openshift/origin/pull/4787#discussion_r40331177>:
>
> > +	if len(args) < 3 {
> > +		return kcmdutil.UsageError(cmd, "POD, LOCAL_SOURCE, and REMOTE_DESTINATION are required arguments")
> > +	}
> > +	if len(args) > 3 {
> > +
> > +	}
> > +	o.PodName = args[0]
> > +	o.Source = args[1]
> > +	o.Destination = args[2]
> > +
> > +	rsh := siblingCommand(cmd, "rsh")
> > +	rshCmd := []string{rsh, o.PodName}
> > +	if len(o.ContainerName) > 0 {
> > +		rshCmd = append(rshCmd, "-c", o.ContainerName)
> > +	}
> > +	rshCmd = append(rshCmd, "sh", "-c", "'exec \"$@\"'")
>
> I'm not using -i in the call to rsh, but are you talking about 'ignore:/'
> in the last parameter to rsync? If I don't have that there, rsync doesn't
> know that it's a remote destination.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787/files#r40331177>.
>

csrwng replied:
thx got it

csrwng replied:
hmm, what if I want to specify a container ? do I still need to do that as a flag on 'oc rsh' ?

smarterclayton replied:
You have to do that as a flag

On Thu, Sep 24, 2015 at 12:48 PM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/rsync.go
> <https://github.com/openshift/origin/pull/4787#discussion_r40342331>:
>
> > +	if len(args) < 3 {
> > +		return kcmdutil.UsageError(cmd, "POD, LOCAL_SOURCE, and REMOTE_DESTINATION are required arguments")
> > +	}
> > +	if len(args) > 3 {
> > +
> > +	}
> > +	o.PodName = args[0]
> > +	o.Source = args[1]
> > +	o.Destination = args[2]
> > +
> > +	rsh := siblingCommand(cmd, "rsh")
> > +	rshCmd := []string{rsh, o.PodName}
> > +	if len(o.ContainerName) > 0 {
> > +		rshCmd = append(rshCmd, "-c", o.ContainerName)
> > +	}
> > +	rshCmd = append(rshCmd, "sh", "-c", "'exec \"$@\"'")
>
> hmm, what if I want to specify a container ? do I still need to do that as
> a flag on 'oc rsh' ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787/files#r40342331>.
>

csrwng commented:
@bparees also suggested adding a --delete flag so that we can delete any files in the destination directory that are not in the source directory. Will work on adding that.

bparees commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,416 @@
+container. If 'rsync' is not present, it will attempt to use 'tar' to send files to the container.`
+
+	rsyncExample = `
+  // Syncrhonize a local directory with a pod directory
```
Synchronize

bparees commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,416 @@
+
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
+	cmd.Flags().StringVar(&o.RsyncExecutable, "rsync", "", "Path to rsync executable")
+	cmd.Flags().StringVar(&o.TarExecutable, "tar", "", "Path to tar executable")
```
these are paths within the container, right?

csrwng replied:
no, paths in your local computer

bparees replied:
so we assume/hope that tar/rsync is in the path on the container?

should we offer a flag for that too?

smarterclayton replied:
I think we should require rsync, tar on your path in both local client and server for now.

csrwng replied:
ok, makes sense. Should I have a flag allowing you to say which one you want to use?

smarterclayton replied:
You should be doing exec.LookPath for both rsync and tar - if rsync isn't found but tar is, print a warning.  If rsync and tar aren't found, print a nice message (which will be different for windows) that says you need rsync on your path.  For windows, we need to point users to a place where they can get an rsync that works.  That's probably https://en.wikipedia.org/wiki/CwRsync - they can grab the free version and extract rsync onto their path somewhere.

bparees replied:
there is no windows rsync that cleanly works, per @csrwng's research, so for windows we'd only support tar.

csrwng replied:
I would only need to look for rsync since we can do tar on the client. 
There's no rsync that works yet on windows without an additional binary to handle the stream redirection:
http://superuser.com/questions/277859/is-there-a-way-to-use-plink-with-rsync-on-windows

csrwng replied:
https://lists.samba.org/archive/rsync/2014-May/029464.html

smarterclayton replied:
I'd like to know the exact problem here - just saying "it seems to be cygwin" isn't sufficient.  We need to be able to actually have an answer for Windows.  Is there someone familiar with cygwin on windows in this company who can answer this question?  Have you looked?

csrwng replied:
I have not looked inside RedHat for an expert. The issue that the cygnative wrapper fixes is described a little bit more here: http://diario.beerensalat.info/2009/03/12/rsync_does_work_with_plink_part_2.html
but of course there's no source available.

csrwng replied:
actually nm, the source is included in the zip download. So as I see it we can:
1) ask you to download this wrapper in addition to cwRsync
2) write code that does something similar ourselves
3) fallback to tar on windows 
preference?

smarterclayton replied:
Can you ask on aos-devel?

On Sep 24, 2015, at 2:21 PM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rsync.go
<https://github.com/openshift/origin/pull/4787#discussion_r40353307>:

> +	}
> +	cmd := &cobra.Command{
> +		Use:     fmt.Sprintf("%s POD LOCAL_SOURCE REMOTE_DESTINATION", name),
> +		Short:   "Copy local files to a pod",
> +		Long:    rsyncLong,
> +		Example: fmt.Sprintf(rsyncExample, fullName),
> +		Run: func(c *cobra.Command, args []string) {
> +			kcmdutil.CheckErr(o.Complete(f, c, args))
> +			kcmdutil.CheckErr(o.Validate())
> +			kcmdutil.CheckErr(o.Run())
> +		},
> +	}
> +
> +	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
> +	cmd.Flags().StringVar(&o.RsyncExecutable, "rsync", "", "Path to rsync executable")
> +	cmd.Flags().StringVar(&o.TarExecutable, "tar", "", "Path to tar executable")

I have not looked inside RedHat for an expert. The issue that the cygnative
wrapper fixes is described a little bit more here:
http://diario.beerensalat.info/2009/03/12/rsync_does_work_with_plink_part_2.html
but of course there's no source available.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787/files#r40353307>.

smarterclayton replied:
I don't really want to maintain Rsync, just want to understand whether the
problem is fixable within our go code (vs the problems encountered via the
Windows shell)

On Sep 24, 2015, at 2:30 PM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rsync.go
<https://github.com/openshift/origin/pull/4787#discussion_r40354519>:

> +	}
> +	cmd := &cobra.Command{
> +		Use:     fmt.Sprintf("%s POD LOCAL_SOURCE REMOTE_DESTINATION", name),
> +		Short:   "Copy local files to a pod",
> +		Long:    rsyncLong,
> +		Example: fmt.Sprintf(rsyncExample, fullName),
> +		Run: func(c *cobra.Command, args []string) {
> +			kcmdutil.CheckErr(o.Complete(f, c, args))
> +			kcmdutil.CheckErr(o.Validate())
> +			kcmdutil.CheckErr(o.Run())
> +		},
> +	}
> +
> +	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
> +	cmd.Flags().StringVar(&o.RsyncExecutable, "rsync", "", "Path to rsync executable")
> +	cmd.Flags().StringVar(&o.TarExecutable, "tar", "", "Path to tar executable")

actually nm, the source is included in the zip download. So as I see it we
can:
1) ask you to download this wrapper in addition to cwRsync
2) write code that does something similar ourselves
3) fallback to tar on windows
preference?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787/files#r40354519>.

smarterclayton commented:
Re: `--delete` - check to see whether rsync already has a name for this.  Also... should we be able to pass arguments to rsync?

liggitt commented:
not sure I like calling it rsync, then behaving differently than rsync. I'd rather call it just "sync" or something and support a couple sane options

csrwng commented:
We already hace rsh

smarterclayton commented:
Sync is reserved for the controller stuff. Rsync is what it *should* be
doing - tar is really a fallback id like to go away.  I would imagine
extending this to be more rsync like in the future rather than less

On Sep 25, 2015, at 9:01 PM, Jordan Liggitt <notifications@github.com>
wrote:

not sure I like calling it rsync, then behaving differently than rsync. I'd
rather call it just "sync" or something and support a couple sane options

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787#issuecomment-143387062>.

csrwng commented:
There is a --delete flag on rsync as well. I would think we can pass flags we don't recognize through to rsync.

smarterclayton commented:
In the future potentially. Definitely trying to get the simple path in
place.  We can add inotify style watches later

On Sep 26, 2015, at 12:58 AM, Cesar Wong <notifications@github.com> wrote:

There is a --delete flag on rsync as well. I would think we can pass flags
we don't recognize through to rsync.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787#issuecomment-143399111>.

csrwng commented:
Updated to use go's tar library for the tar fallback, addressed other comments above. However, there's still some issues with tar on windows. Working on a fix.

csrwng commented:
Still requires an update from s2i --> https://github.com/openshift/source-to-image/pull/295
Now using a temporary file for tar instead of a pipe. Writing to a pipeline writer in windows will not block until a reader is available.

csrwng commented:
ok, this should be ready for a final review. In windows, simply opening files with standard descriptors is not yet doing the trick for rsync. Needs further investigation and a follow-up pull. 

@smarterclayton @bparees

smarterclayton commented on pkg/cmd/cli/cmd/rsh.go:
```diff
@@ -28,87 +30,107 @@ directly.`
+	}
+
+	if len(args) < 1 {
+		kcmdutil.UsageError(cmd, "rsh requires a single Pod to connect to")
```
Return

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,372 @@
+
+	rsyncExample = `
+  # Synchronize a local directory with a pod directory
+  $ %[1]s POD ./local/dir/ /remote/dir`
```
Hrm, I prefer exact same syntax as rsync here

smarterclayton replied:
Won't rsync work in both directions?

csrwng replied:
So 'oc rsync ./local/dir [POD]:/remote/dir' ? I like that better

Hmm, and yes, I suppose we could also allow 
'oc rsync [POD]:/remote/dir ./local/dir', 
My understanding is that initially we only wanted client -> pod though.

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,372 @@
+	// Use tar if running on windows
+	if runtime.GOOS == "windows" {
+		o.UseTar = true
+	}
```
Todo about figuring out windows io

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,372 @@
+	}
+
+	return fmt.Errorf("the container in pod %s/%s does not have rsync or tar", o.Namespace, o.PodName)
+}
```
Show Container name, don't show namespace name

csrwng replied:
We need an additional call to figure out what the container name is, is that ok?

smarterclayton replied:
Yeah, even though we will have to parse the arguments in the rsync pattern,
it's more correct and opens up future scenarios.  And it's OK to fetch the
pod info.

On Oct 2, 2015, at 1:07 AM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rsync.go
<https://github.com/openshift/origin/pull/4787#discussion_r40977859>:

> +		// If rsync is present in the pod's container, then use it
> +		if o.checkRemoteRsync() == nil {
> +			return o.copyWithRsync()
> +		}
> +	}
> +
> +	// If tar is present in the pod's container, then use it
> +	if o.checkRemoteTar() == nil {
> +		if !o.UseTar {
> +			glog.Warningf("Could not find rsync. It must be on the path of the local machine and of the remote container, using tar.")
> +		}
> +		return o.copyWithTar()
> +	}
> +
> +	return fmt.Errorf("the container in pod %s/%s does not have rsync or tar", o.Namespace, o.PodName)
> +}

We need an additional call to figure out what the container name is, is
that ok?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787/files#r40977859>.

smarterclayton commented:
Needs more tests - tar path and rsync path.  I'd like rsync path in e2e

smarterclayton commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,372 @@
+}
+
+func (o *RsyncOptions) checkRemoteRsync() error {
+	err := executeWithLogging(o.RemoteExecutor, testRsyncCommand)
```
I'm not a fan of this check - three execs vs one is expensive.  What's the time it adds to the call.

csrwng replied:
It takes 100ms on average. We could also check the return code (14 is IPC error) and error output of the local rsync call. If we can determine that it's because rsync is not present in the container, then fallback to tar.

smarterclayton replied:
Plus latency, so up to 300ms for long distance.  Can't we simply look at
stderr on fallback?  I'd prefer to assume rsync in the container and walk
back to something useful.  Or if the command fails, only then check rsync.

On Oct 2, 2015, at 1:39 AM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rsync.go
<https://github.com/openshift/origin/pull/4787#discussion_r40979787>:

> +	err := e.Execute(cmd, nil, w, w)
> +	glog.V(4).Infof("%s", w.String())
> +	glog.V(4).Infof("error: %v", err)
> +	return err
> +}
> +
> +func (o *RsyncOptions) checkLocalRsync() error {
> +	_, err := exec.LookPath("rsync")
> +	if err != nil {
> +		glog.V(3).Infof("local rsync not found: %v", err)
> +	}
> +	return err
> +}
> +
> +func (o *RsyncOptions) checkRemoteRsync() error {
> +	err := executeWithLogging(o.RemoteExecutor, testRsyncCommand)

It takes 100ms on average. We could also check the return code (14 is IPC
error) and error output of the local rsync call. If we can determine that
it's because rsync is not present in the container, then fallback to tar.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787/files#r40979787>.

bparees commented:
lgtm aside from needing tests.

bparees commented:
you might talk to @jhadvig and @mnagy about incorporating oc sync testing into the hot reload tests they are writing.

mnagy commented:
@bparees @csrwng incorporating it into our tests shouldn't be a problem. Right now we're using `oc exec` with `sed` to modify the source code.

csrwng commented:
[test]

csrwng commented:
[test]

csrwng commented:
Added extended test and e2e test

csrwng commented:
@mfojtik

mfojtik commented:
@csrwng do you want to retag this to run only extended and your test?

mfojtik commented on test/extended/devtools/rsync.go:
```diff
@@ -0,0 +1,64 @@
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Verifying that the expected files are in the container")
+			result, err = oc.Run("rsh").Args(podName, "ls", "/tmp/image-streams").Output()
```
can you also write some error case? Like copying the file into directory where you don't have permissions to write (and verify that you get good error back)

mfojtik commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,411 @@
+// defaultLocalExecutor will execute commands on the local machine
+type defaultLocalExecutor struct{}
+
+func (*defaultLocalExecutor) Execute(command []string, in io.Reader, out, errOut io.Writer) error {
```
godoc

mfojtik commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,411 @@
+	Config        *kclient.Config
+}
+
+func (e *defaultRemoteExecutor) Execute(command []string, in io.Reader, out, errOut io.Writer) error {
```
godoc

mfojtik commented on pkg/cmd/cli/cmd/rsync.go:
```diff
@@ -0,0 +1,411 @@
+	// Use tar if running on windows
+	// TODO: Figure out how to use rsync in windows so that I/O can be
+	// redirected from the openshift native command to the cygwin rsync command
+	if runtime.GOOS == "windows" {
```
there should be some constant for "windows" ;-) just a nit I can live with this as well :)

mfojtik commented:
@csrwng since you're bumping S2I, can you please include @jhadvig commit? https://github.com/openshift/source-to-image/commit/1fd4429c584d688d83c1247c03fa2eeb0b083ccb

mfojtik commented:
small nits, but LGTM

just wonder why you use "devtools" pkg for extended test and not "cli" (we might want to add more tests for CLI commands that we don't usually test, or more error cases we also don't test that much in test-cmd.sh).

also please tag this to run the extended test, to make sure it pass ;-)

csrwng commented:
cli sounds much better, will change

csrwng commented:
@mfojtik thx. I updated the s2i bump to the latest commit and addressed comments. 
[testonlyextended][extended:core(gitauth)]

mfojtik commented:
LGTM

csrwng commented:
ha, just realized I tagged the wrong test!
[testonlyextended][extended:core(rsync)]

@smarterclayton if you get a chance, please take a look as well. Your comments should be addressed.

bparees commented:
lgtm [merge]

if @smarterclayton has further concerns they can be handled in a follow up.

smarterclayton commented:
Can we make sure we have a card to cover the Windows STDIN issues?  I want
rsync on windows to not get forgotten.

On Fri, Oct 9, 2015 at 1:16 PM, Ben Parees <notifications@github.com> wrote:

> lgtm [merge]
>
> if @smarterclayton <https://github.com/smarterclayton> has further
> concerns they can be handled in a follow up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787#issuecomment-146936365>.
>

csrwng commented:
Trello card here: https://trello.com/c/bsAQVg0E/718-support-windows-rsync-with-oc-rsync

smarterclayton commented:
Do you have a plan to resolve this before 3.1?

On Fri, Oct 9, 2015 at 3:22 PM, Cesar Wong <notifications@github.com> wrote:

> Trello card here:
> https://trello.com/c/bsAQVg0E/718-support-windows-rsync-with-oc-rsync
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787#issuecomment-146966029>.
>

csrwng commented:
@smarterclayton I did a little bit more digging in this area. The magic that the cygnative code does is that it's using both native win32 library i/o and i/o functions provided by cygwin. It pipes from one to the other. Usually it's either/or for a program. Go by default is using the win32 native calls. It's really not hard to dynamically load the cygwin1.dll and invoke functions on it from go (https://github.com/golang/go/wiki/WindowsDLLs). 

However, the issue is what we're willing to require for our code to function. The free version of cwRsync comes only with the 32-bit version of cygwin1.dll. In order for us to load it and use it, our go code needs to be compiled for i386. Alternatively, we could require you to download the 64-bit version of cygwin and add it to your path. 

So we have the following options:
1) Write a separate binary that's 32-bit windows that will basically perform the function of cygnative. This will only require you to download the free version of cwRsync and you're set. Pros - less requirement on you as a user. Cons - more headache for us to carry a separate binary.

1b- ish) Simply say the separate binary is the cygnative exe and we tell you where to download it. Pros: can be done now, Cons: really bad experience. Not sure if we can do it legally.

2) Embed the code that pipes from one library to the other in our 64-bit binary and require you to download cygwin 64-bit and add it to your path. Pros: no additional binaries for us to package. Cons: more work for our users, more chance for things to not be setup properly.

3) Make our windows distribution 32-bit only instead of 64-bit? And do something like #2 above. Not sure if this can even be considered.

smarterclayton commented:
Why do we need to use cygwin1.dll?  Why does #2 require cgywin 64 bit?
Pipes are across process boundaries and shouldn't' be specific to arch of
whatever cgywin.

On Tue, Oct 13, 2015 at 9:21 AM, Cesar Wong <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> I did a little bit
> more digging in this area. The magic that the cygnative code does is that
> it's using both native win32 library i/o and i/o functions provided by
> cygwin. It pipes from one to the other. Usually it's either/or for a
> program. Go by default is using the win32 native calls. It's really not
> hard to dynamically load the cygwin1.dll and invoke functions on it from go
> (https://github.com/golang/go/wiki/WindowsDLLs).
>
> However, the issue is what we're willing to require for our code to
> function. The free version of cwRsync comes only with the 32-bit version of
> cygwin1.dll. In order for us to load it and use it, our go code needs to be
> compiled for i386. Alternatively, we could require you to download the
> 64-bit version of cygwin and add it to your path.
>
> So we have the following options:
> 1) Write a separate binary that's 32-bit windows that will basically
> perform the function of cygnative. This will only require you to download
> the free version of cwRsync and you're set. Pros - less requirement on you
> as a user. Cons - more headache for us to carry a separate binary.
>
> 1b- ish) Simply say the separate binary is the cygnative exe and we tell
> you where to download it. Pros: can be done now, Cons: really bad
> experience. Not sure if we can do it legally.
>
> 2) Embed the code that pipes from one library to the other in our 64-bit
> binary and require you to download cygwin 64-bit and add it to your path.
> Pros: no additional binaries for us to package. Cons: more work for our
> users, more chance for things to not be setup properly.
>
> 3) Make our windows distribution 32-bit only instead of 64-bit? And do
> something like #2 <https://github.com/openshift/origin/pull/2> above. Not
> sure if this can even be considered.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787#issuecomment-147712069>.
>

csrwng commented:
For a go-lang app to use a dll it needs to match the architecture of the dll. Yes, a 64-bit process can spawn a 32-bit and vice versa and they can pipe i/o. That's not the issue. You need the cygwin1.dll because the way that cygwin handles console I/O is very different than native windows. We're trying to pipe I/O from a native cygwin program (cwrsync) to a native windows program.

smarterclayton commented:
It's just bytes on a file descriptor?  Can you describe the special format
that cygwin is providing IO in?

On Tue, Oct 13, 2015 at 11:15 AM, Cesar Wong <notifications@github.com>
wrote:

> For a go-lang app to use a dll it needs to match the architecture of the
> dll. Yes, a 64-bit process can spawn a 32-bit and vice versa and they can
> pipe i/o. That's not the issue. You need the cygwin1.dll because the way
> that cygwin handles console I/O is very different than native windows.
> We're trying to pipe I/O from a native cygwin program (cwrsync) to a native
> windows program.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787#issuecomment-147745940>.
>

csrwng commented:
I'm researching to give you an intelligent answer :-) I don't believe it is
the format of the I/O, but rather how handle inheritance is setup from the
calling process (the cygwin rsync) to its child process (our oc). Apart
from cygwin source, there's very little to go on in terms of
doc/discussions.

So far I only understand what cygnative is doing to work around the problem
and that's to use the cygwin dll to work with the streams from the cygwin
app.

On Tue, Oct 13, 2015 at 1:45 PM, Clayton Coleman <notifications@github.com>
wrote:

> It's just bytes on a file descriptor? Can you describe the special format
> that cygwin is providing IO in?
>
> On Tue, Oct 13, 2015 at 11:15 AM, Cesar Wong <notifications@github.com>
> wrote:
>
> > For a go-lang app to use a dll it needs to match the architecture of the
> > dll. Yes, a 64-bit process can spawn a 32-bit and vice versa and they can
> > pipe i/o. That's not the issue. You need the cygwin1.dll because the way
> > that cygwin handles console I/O is very different than native windows.
> > We're trying to pipe I/O from a native cygwin program (cwrsync) to a
> native
> > windows program.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/4787#issuecomment-147745940>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787#issuecomment-147791008>.
>

smarterclayton commented:
But cgywin rsync is sending it to cygwin native in exactly the same way.
So the way cygwin native reads it has nothing to do with the architecture
of cygwin rsync.

On Tue, Oct 13, 2015 at 2:40 PM, Cesar Wong <notifications@github.com>
wrote:

> I'm researching to give you an intelligent answer :-) I don't believe it is
> the format of the I/O, but rather how handle inheritance is setup from the
> calling process (the cygwin rsync) to its child process (our oc). Apart
> from cygwin source, there's very little to go on in terms of
> doc/discussions.
>
> So far I only understand what cygnative is doing to work around the problem
> and that's to use the cygwin dll to work with the streams from the cygwin
> app.
>
> On Tue, Oct 13, 2015 at 1:45 PM, Clayton Coleman <notifications@github.com
> >
> wrote:
>
> > It's just bytes on a file descriptor? Can you describe the special format
> > that cygwin is providing IO in?
> >
> > On Tue, Oct 13, 2015 at 11:15 AM, Cesar Wong <notifications@github.com>
> > wrote:
> >
> > > For a go-lang app to use a dll it needs to match the architecture of
> the
> > > dll. Yes, a 64-bit process can spawn a 32-bit and vice versa and they
> can
> > > pipe i/o. That's not the issue. You need the cygwin1.dll because the
> way
> > > that cygwin handles console I/O is very different than native windows.
> > > We're trying to pipe I/O from a native cygwin program (cwrsync) to a
> > native
> > > windows program.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/openshift/origin/pull/4787#issuecomment-147745940
> >.
> >
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/4787#issuecomment-147791008>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4787#issuecomment-147807292>.
>

csrwng commented:
@smarterclayton I chatted with someone from the cygwin list yesterday and he explained that cygwin fd's != Windows file handles. They point to different things in memory that do not interop. It is made worse in this case because cwrsync (the parent process) is creating an anonymous pipe and using its fds to replace standard fds on the child process's file table. Only cygwin knows how to write to it...  native windows APIs can't work with its in-memory representation.

csrwng commented:
@smarterclayton Just wanted to run an idea by you that wouldn't require us to try to find a cygwin dll. Basically we use rsh to start an rsync daemon on the pod container, use port-forward to forward it to a local port on the windows machine and then tell rsync to use the rsync protocol on the local port. Do you think it would be worth pursuing?

smarterclayton commented:
Good thought.  Two exec sessions def more expensive than one, but if it
works...

On Oct 16, 2015, at 10:08 AM, Cesar Wong <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> Just wanted to run an
idea by you that wouldn't require us to try to find a cygwin dll. Basically
we use rsh to start an rsync daemon on the pod container, use port-forward
to forward it to a local port on the windows machine and then tell rsync to
use the rsync protocol on the local port. Do you think it would be worth
pursuing?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4787#issuecomment-148726884>.
